src/ros_commo.py

```diff
@@ -376,15 +402,85 @@ def behavior_switch_callback(self, data):
 	# Data is a bit-flag that enables/disables publication of messages.
 	def behavior_control_callback(self, data):
 		self.control_mode = data.data
-
+	
+	def face_loc_cb(self, data):
+		#print "@@@@@@@@@@ In face cb"
+		self.lock.acquire()
+		try:
+			if len(self.sc_str_set)<1:
+				self.face_array=data.faces
+				self.sc_str_set="do"
+		finally:
+			self.lock.release()
+		#for face in data.faces:
+			#fac="(map-ato \"faces\" (NumberNode \""+str(face.id)+"\") "+str(face.point.x)+" "+str(face.point.y)+" "+str(face.point.z)+")"
+			#scheme_eval(self.atomspace,fac)
+
+	def thr(self):
+		self.atomspace = AtomSpace()#scheme_eval_as('(cog-atomspace)')
+		lpth="/home/mandeep/hr/opencog/ros-behavior-scripting/src/time-map.scm"
+		scheme_eval(self.atomspace, "(load \""+lpth+"\")")
+		#print scheme_eval(self.atomspace,"(+ 2 3)")
+		#scheme_eval(self.atomspace, "(map-ato \"faces\" (NumberNode \"2.0\") 1 2 3)")
+		rate=rospy.Rate(30)
+		while not rospy.is_shutdown():
+			rate.sleep()
+			self.lock.acquire()
+			try:
+				if len(self.sc_str_set)>0:
+					#print "faces: ", len(self.face_array)
+					for face in self.face_array:
+						#print "hello"
+						fac="(map-ato \"faces\" (NumberNode \""+str(face.id)+"\") "+str(face.point.x)+" "+str(face.point.y)+" "+str(face.point.z)+")"
+						scheme_eval(self.atomspace,fac)
+					self.sc_str_set=""
```

Comment:
why are you blasting this into the atomspace at 30 frames a second?

Instead of making this ros node more complicated, a much better design would be to create a completely new ros node, and have that node subscribe to the face location messages. Then whenever a new face location shows up, ROS will automatically call your callback, and then, from that callback, you can poke the new 3D location values into  the atomspace.

This is kind of the whole point of having ROS nodes to begin with: they make your code modular and distributed and easy to maintain. They avoid any need for threads -- python is just not multi-threaded.   That means you can avoid writing spin-loops as you did here, and you can avoid using locks.  In addition, your code will run faster on a multi-CPU machine, since each ROS node can run on its own CPU.   By cnotrast, python is single-threaded, which means all python code must always run on exactly one CPU, and thus can never eve get any speedup on a multi-cpu machine.

This is the reason that the concept of ROS nodes exist. -- use ROS nodes and the publish-subscribe model that it provides. Do NOT use threads and spin-loops!